querybook/server/logic/generic_permission.py

```diff
@@ -0,0 +1,64 @@
+from typing import List, Optional, Tuple, Union
+from app.db import with_session
+from models import UserGroupMember, User, DataDocEditor, BoardEditor
+from sqlalchemy import func, select
+from sqlalchemy.orm import Session
+
+
+@with_session
+def get_all_groups_and_group_members_with_access(
+    doc_or_board_id: int,
+    editor_type: Union[DataDocEditor, BoardEditor],
+    uid: Optional[int] = None,
+    session: Optional[Session] = None,
+) -> List[Tuple[int, int, bool, bool]]:
+    """
+    Get all groups and group members with access to a DataDoc or Board.
+
+    Args:
+        doc_or_board_id: The ID of the DataDoc or Board.
+        editor_type: The type of editor (DataDocEditor or BoardEditor).
+        uid: The ID of the user to filter by (optional)
+        session: The database session to use.
+
+    Returns:
+        A list of tuples containing the editor ID, the group or user ID, and the most permissive read and write permissions.
+        Editors with inherited permissions have their ID set to None.
+    """
+    topq = session.query(
+        editor_type.id, editor_type.uid, editor_type.read, editor_type.write
+    ).select_from(editor_type)
+
+    if editor_type == DataDocEditor:
+        topq = topq.filter(editor_type.data_doc_id == doc_or_board_id)
+    elif editor_type == BoardEditor:
+        topq = topq.filter(editor_type.board_id == doc_or_board_id)
+
+    topq = topq.cte("cte", recursive=True)
+
+    bottomq = (
+        select([None, UserGroupMember.uid, topq.c.read, topq.c.write])
+        .select_from(topq)
+        .join(User, topq.c.uid == User.id)
+        .join(UserGroupMember, UserGroupMember.gid == User.id)
+        .filter(User.is_group)
+    )
+
+    recursive_q = topq.union(bottomq)
+
+    editors = recursive_q.alias()
+
+    q = select(
+        [
+            func.max(editors.c.id),
+            editors.c.uid,
+            func.max(editors.c.read),
+            func.max(editors.c.write),
```

Comment:
We tested this with Postgres and it failed (as you cannot use the `max` function with a boolean). We eluded this by casting the field to `Integer`, like so:
```suggestion
            func.max(editors.c.read.Integer),
            func.max(editors.c.write.Integer),
```

Reply:
Hi @anacltd, I'm getting an error doing it that way but I was able to work around it by adding 
`func.max(editors.c.read.cast(Integer))` instead of `func.max(editors.c.read.Integer)`. Does this way work for you?

Reply:
so I also get an error when working locally with a MySQL DB, but in production we use a Postgres DB (with which is seems to work just fine). I think the `cast` function is a good way to make it work on both, so kudos for finding this!